---
title: "CPS1000_Analysis"
author: ""
date: "`r doc_date()`"
output:
  BiocStyle::html_document:
      toc: yes
      toc_depth: 3
      toc_float: yes
      code_folding: "hide" 
editor_options: 
  chunk_output_type: inline
---

This R code includes the analyses and figures of the CPS1000 manuscript, written by Kevin D. Hofer and Junyan Lu.

# Setup
```{r options, message=FALSE, warning = FALSE}
library(knitr)
options(digits=3, width=80)
opts_chunk$set(echo=TRUE,tidy=FALSE,include=TRUE)
```

## Load libraries
```{r Load packages, message=FALSE, warning=FALSE}
library(ggpmisc)
suppressMessages(library(RColorBrewer))
suppressMessages(library(ggrepel))
suppressMessages(library(cowplot))
suppressMessages(library(gridExtra))
suppressMessages(library(drc))
suppressMessages(library(tidyverse))
detach("package:dplyr") 
suppressMessages(library(dplyr))
library("survminer")
library("survival")
library("maxstat")
library("patchwork")
library("glmnet")
library("corrplot")
library("writexl")
library("ggplot2")
library("flextable")
library("ggpubr")
library("ComplexHeatmap")
library("circlize")
library("grid")
library("RColorBrewer")
library("ggplotify")
library("magrittr")
library("Rtsne")
library("paletteer")
library("ggtext")
library("ggh4x")
library("plotrix")
library("scales")
library("corrplot")
library("gridGraphics")
library("draw")
library("stringr")


#maintain function of specific R packages
rename <- dplyr::rename
count  <- dplyr::count
filter <- dplyr::filter
```

## Define figure themes
```{r themes}

theme_figures <- theme_classic()+
  theme(plot.title = element_text(size=8, hjust=0.5, face="bold"),
        axis.text = element_text(color="black", size=8), 
        axis.title = element_text(size=8), 
        axis.line = element_line(size=0.5), 
        legend.title = element_text(size=8, face="bold"), 
        legend.text = element_text(size=8), 
        legend.key.size = unit(0.5, 'cm'))

theme_figures_angle45_x <- theme_classic()+
  theme(plot.title=element_text(size=8, hjust=0.5, face="bold"),
        axis.text.x = element_text(color="black", angle = 45, hjust=1, vjust=1, size=8), 
        axis.text.y = element_text(size=8), 
        axis.title = element_text(size=8), 
        legend.title = element_text(size=8, face="bold"), 
        legend.text = element_text(size=8), 
        axis.line = element_line(size=0.5), 
        legend.key.size = unit(0.5, 'cm'))

theme_figures_facet <- theme_bw() +   
  theme(strip.text.y = element_text(size = 0), 
        strip.text.x = element_text(size=8),
        strip.background.y = element_rect(color=NA),
        axis.text = element_text(size=8, color="black"),
        axis.title = element_text(size=8, color="black"),
        panel.spacing = unit(0.15, "lines"), 
        legend.position = "none",
        panel.border = element_rect(color = "black", linewidth = 0.75, fill = NA),
        panel.grid = element_blank(),
        strip.background = element_rect(color = "black", linewidth = 0.75, fill = "grey90"),
        plot.title = element_text(hjust=0.5, size=8, face="bold"))

theme_figures_facet_angle60_x <- theme_bw() +   
  theme(strip.text.y = element_text(size = 0), 
        strip.text.x = element_text(size=8),
        strip.background.y = element_rect(color=NA),
        axis.text.y = element_text(size =8, color="black"),
        axis.text.x = element_text(angle = 60, hjust = 1, vjust = 1, size=8, color = "black"),
        panel.spacing = unit(0.15, "lines"), 
        legend.title = element_text(size = 8),
        panel.border = element_rect(color = "black", linewidth = 0.75, fill = NA),
        strip.background = element_rect(color = "black", linewidth = 0.75, fill = "grey90"),
        plot.title = element_text(hjust=0.5, size=8, face="bold"), 
        strip.text = element_text(size = 8))

#set colors
colors_diag <- setNames(c("#1F78B4", "#B2DF8A", "#A6CEE3", "grey", "#E31A1C", "#FB9A99", "#FF7F00", "#CAB2D6"), c("MCL", "CLL", "T-PLL", "other", "AML", "T-ALL", "B-ALL", "B-PLL"))

colors_diag_IGHV <- setNames(c("#1F78B4", "#B2DF8A", "#33A02C", "grey80", "#E31A1C", "#FB9A99", "#FF7F00", "#CAB2D6", "#A6CEE3"), c("MCL", "U-CLL", "M-CLL", "other", "AML", "T-ALL", "B-ALL", "B-PLL", "T-PLL"))

colors_pathways_mod <- setNames(c("lightgrey", "#BC80BD", "#FCCDE5", "#B3DE69", "#FDB462", "#80B1D3", "#FB8072", "#BEBADA", "#FFFFB3", "#8DD3C7"), c("other", "AKT/mTOR", "DNA damage response", "Chemotherapy" ,  "MAPK" , "B-cell receptor", "PI3K", "Stress pathway", "Bromodomain/PLK", "JAK/STAT"))

colors_cor_heatmap <- setNames(c("#BC80BD", "#FDB462", "#80B1D3", "#FB8072", "#BEBADA"), c("AKT/mTOR", "MEK/ERK" , "B-cell receptor", "PI3K", "Stress pathway"))

colors_drug_type <- c("Chemo" = "#9a6a3d", "Kinase" = "#6A3D9A", "Other" = "#3d9a6a")

```

## Define functions
### Functions for supplementary tables
```{r}

stab_theme <- function(ft) {
  ft |>
    theme_alafoli() |>
    autofit() |> 
    fontsize(size = 8, part = "header") |> 
    fontsize(size = 8, part = "body") |> 
    padding(padding.top = 3, padding.bottom = 3, part = "all") |> 
    align(align = "left", part = "body") |>  
    align(align = "left", part = "header") |>
    bold(part = "header") |>
    width(j = 1, width = 1.25) |>
    width(j = 2, width = 3.5) |>
    width(j = 3, width = 0.75) |>
    line_spacing(space = 1, part = "all") |> 
    color(color = "black", part = "all")
}
```

### Functions for analyses (???)
```{r}

#function for t-test
tTest <- function(val, fac) {
  res <- t.test(val ~ fac, var.equal = TRUE)
  data.frame(p = res$p.value, 
             mean1 = res$estimate[[1]],
             mean2 = res$estimate[[2]],
             diff = res$estimate[[2]] - res$estimate[[1]])
}

#helper function for mean viability boxplots
create_boxplot <- function(data, diagnosis_filter, colors, title) {
  #order drugs by median viability
  drug_order <- data |> 
    filter(diagnosis == diagnosis_filter) |> 
    group_by(drug) |> 
    summarise(median = median(mean_viab)) |> 
    arrange(desc(median)) |> 
    pull(drug)
  
  #filter and reorder data
  plot_data <- data |> 
    filter(diagnosis == diagnosis_filter) |> 
    mutate(drug = factor(drug, levels = drug_order))
  
  #create plot
  plot_data |> 
    filter(diagnosis == diagnosis_filter) |> 
    ggplot(aes(x=mean_viab, y=drug, color=diagnosis))+ 
    geom_boxplot(alpha=0.8, outlier.shape=NA) +
    geom_jitter(alpha=0.2, size=0.2, width = 0.15)+
    scale_x_continuous(name ="Mean viability", breaks=seq(0,1.2, 0.2), limits=c(0,1.2))+
    theme_bw() +
    theme(
      axis.text = element_text(color = "black", size = 8),
      axis.title = element_text(size = 8),
      plot.title = element_text(size = 8, hjust = 0.5, face = "bold"),
      legend.position = "none",
      panel.border = element_rect(color = "black", linewidth = 1, fill = NA)) +
    labs(y = "", x = "Mean viability", title = title) +
    scale_color_manual(values = colors_diag_IGHV)
}

```

# Load data
## File path
```{r filepath} 
# Modify according to directory

#filepath <-"../../"
filepath <- "~/Dropbox/Medizin/Forschung/Aktuelle Forschungsprojekte/CLL CPS1000/figures/github/"

```

## Load processed CPS1000 data
```{r}
load(paste0(filepath,"data/CPS1000_data.RData"))
```
Description of datasets:  
- screen: processed DMSO-normalized ex vivo viability measurements for all patients and drugs
- drugs: list of drugs used in the screen
- patients: list of patients with corresponding diagnoses
- metadata: annotation of genetic aberrations
- treatments: treatment contexts for the longitudinal analysis of paired samples
- survival: overall survival and time-to-treatment data

# Drugs
## Drug overview (Fig. 2A)
```{r, fig.width = 12, fig.height = 6}
#prepare data
drugs <- drugs |> as.data.frame()
as.factor(drugs$Pathway)

Fig2a <- ggplot(drugs, aes(x = factor(Pathway))) + 
geom_bar(aes(fill = Type)) + 
scale_fill_brewer(palette = "PuRd",  direction = 1) +
labs(y="Number of drugs", x="", fill="Drug type") +
theme_figures_angle45_x +
scale_y_continuous(expand = c(0.01,0), breaks=c(1,2,3,4,5,6,7,8)) +
scale_x_discrete(expand = c(0.025,0)) +
theme(plot.margin = margin(t=1, b=1, r=1, l=1, unit="cm"))
Fig2a

ggsave(path=(paste0(filepath,"figures")), filename = "Fig2a_drug_compounds.svg")

```


## Drug list
```{r}

#prepare data
drugs$Drug <- drugs$Drug |> 
  gsub("\\.", "-", x = _) |>
  gsub("D4476", "D 4476", x = _) |>
  gsub("IRAK1/4 Inhibitor I", "IRAK-1/4 Inhibitor I", x = _) |> 
  gsub("SB-203580", "SB 203580", x = _)

drugs$Drug <- sort(drugs$Drug) 
drug_list <- drugs |> 
  dplyr::select(Drug, Target, Pathway) #add supplier

drugs$Drug

stab3 <- flextable(drug_list) # |> stab_theme()

ft_grob <- gen_grob(stab3, fit = "width", scaling = "min", hjust = 0)

STab3 <- ggplot() +
  annotation_custom(ft_grob) +
  theme_void() +
  theme(plot.margin = margin(t = 1, r = 1, b = 1, l = 1, unit = "cm"))

ggsave(width=21, height=29.7, units = "cm", path=(paste0(filepath,"tables")), filename = "STab3_drugs_and_targets.svg")
```

# Patient characteristics
## Diagnoses (Fig. 2C)
```{r, fig.width = 12, fig.height = 6}

#prepare data
patients <- patients |> 
  filter(diagnosis != "") |> 
  as.data.frame()

patients$cohort <- patients$cohort |> 
  str_replace("IC50", "IC50 + CPS1000") |>
  as.factor()
  
Fig2c <- ggplot(patients, aes(x = forcats::fct_infreq(diagnosis))) +
  geom_bar(aes(fill = cohort)) + 
  geom_text(stat = "count", aes(label = after_stat(count)), vjust = -0.5, size=2.5) +
  scale_fill_brewer(palette = "Greens", direction =-1)+
  labs(y="Number of patients", x="", fill="Drug screen")+
  theme_figures_angle45_x +
  scale_y_continuous(expand = c(0.02,0), breaks=seq(0,300,50), limits = c(0,300))+
  scale_x_discrete(expand = c(0.04,0)) +
  theme(plot.margin = margin(t=1, b=1, r=1, l=1, unit="cm"))
Fig2c

ggsave(path=(paste0(filepath,"figures")), filename = "Fig2c_disease_distribution.svg")

```

## Longitudinal data (Fig. 2D)
```{r, fig.width = 3, fig.height = 6}

#prepare data
treatments$sample1 <- gsub("-[1234]$", "", treatments$sample1)
treatments$sample2 <- gsub("-[1234]$", "", treatments$sample2)

length(unique(treatments$patientID))

#merge data
sample_date <- survival[,1:3]
merged_table <- merge(treatments, sample_date, by.x="sample1", by.y="sampleID", all.x=TRUE) |> 
  merge(patients[,c("patientID", "diagnosis")], by="patientID") |> 
  dplyr::rename(sampleDate1 = sampleDate)

#add date for 2. sampling and calculate time interval
merged_table <- merge(sample_date, merged_table, by.x="sampleID", by.y="sample2", ) |> 
  dplyr::rename(sample2 = sampleID, sampleDate2 = sampleDate) |> 
  dplyr::select(c(patientID, sample1, sample2, sampleDate1, sampleDate2, diagnosis)) |> 
  mutate(diff = sampleDate2-sampleDate1)

#plot
merged_table$diff <- as.numeric(merged_table$diff)
summary(merged_table$diff)

Fig2d <- ggplot(merged_table, aes(x="", y=diff)) +
  geom_boxplot(color="black", linewidth=0.5, outliers = FALSE, width=0.4) +
  geom_jitter(width = 0.1, alpha = 0.75, size=1, color="#74c476")+ 
  theme_figures+
  labs(x = "", y = "Days between samples")+
  theme(axis.ticks.x = element_blank())+
  scale_y_continuous(expand = c(0.02,0), breaks=c(seq(0,1750,500))) +
  theme(plot.margin = margin(t=1, r=1, l=1, b=1, unit="cm"))
Fig2d

ggsave(path=(paste0(filepath,"figures")), filename = "Fig2d_interval_samples.svg")

```

## Oncoprint of genetic alterations in the CLL cohort (Fig. 2E)
```{r, fig.width = 12, fig.height = 8}

#prepare data
#remove del5IgH
oncoprint_data <- merge(patients, metadata, by.x="patientID", by.y="Patient.ID") |> 
  filter(diagnosis.x == "CLL") |> 
  dplyr::select(!del5IgH)

#remove the first two columns (IGHV.status and Methylation_Cluster)
matrix_data <- as.matrix(oncoprint_data[, -c(1:13)])
rownames(matrix_data) <- oncoprint_data$patientID 
genetic_variations <- t(matrix_data)

#exclude aberrations with less than 8 events
genetic_variations <- data.matrix(genetic_variations)

genetic_variations_num <- matrix(as.numeric(genetic_variations), ncol = 275)
rownames(genetic_variations_num) <- rownames(genetic_variations)
colnames(genetic_variations_num) <- colnames(genetic_variations)

genetic_variations_num <- as.data.frame(genetic_variations_num)
genetic_variations_num$count <- rowSums(!(is.na(genetic_variations_num) | genetic_variations_num == 0)) 

genetic_variations_num <- genetic_variations_num |> 
  filter(count >7)
  
#re-adjust matrix für heatmap
genetic_variations_num[genetic_variations_num==0]<-""
genetic_variations_num[is.na(genetic_variations_num)]<-""

genetic_variations_red <- dplyr::select(genetic_variations_num, !count)
genetic_variations_red <- as.matrix(genetic_variations_red)

#annotations
#annotation for IGHV status
ighv_anno <- oncoprint_data[,"IGHV.status"]
ighv_anno <- ighv_anno %>% replace(is.na(.), "unknown")
ighv_colors <- colorRampPalette(brewer.pal(9, "Set1"))(length(unique(ighv_anno)))
annotation_colors <- setNames(ighv_colors, unique(ighv_anno))

#annotation for Methylation
meth_anno <- oncoprint_data[,"Methylation_Cluster"]
meth_anno <- factor(meth_anno, levels = c("HP", "IP","LP", "unknown"))
meth_anno <- meth_anno %>% replace(is.na(.), "unknown")
meth_colors <- colorRampPalette(brewer.pal(8, "Set2"))(length(unique(meth_anno)))
annotation_colors_meth <- setNames(meth_colors, unique(meth_anno))

#annotation for center
center_anno <- oncoprint_data[,"project"]
center_anno <- factor(center_anno, levels = c("HD", "HD-DNA", "H68", "MIR34a", "FB", "NOV", "MDACC", "MUN", "HCL_HD_Andrulis", "OX", "BARC", "ZCH_2"))
center_colors <- colorRampPalette(brewer.pal(3, "Paired"))(length(unique(center_anno)))

#change center label
center_anno <- fct_recode(center_anno,
"Zürich" = "ZCH_2",
"Barcelona" = "BARC",
"Heidelberg" = "HD")
annotation_colors_center <- setNames(center_colors, unique(center_anno))

#change IGHV label
ighv_anno <- fct_recode(ighv_anno,
"mutated" = "M",
"unmutated" = "U")
annotation_colors <- setNames(ighv_colors, unique(ighv_anno))

ha <- HeatmapAnnotation(Center = center_anno, IGHV = ighv_anno, Methylation = meth_anno, annotation_name_gp = gpar(fontsize = 8), col = list(Center =  annotation_colors_center, IGHV = annotation_colors, Methylation = annotation_colors_meth), annotation_legend_param = list(
    Center = list(
      title_gp = gpar(fontsize = 8, fontface = "bold"),
      labels_gp = gpar(fontsize = 8)
    ),
    IGHV = list(
      title_gp = gpar(fontsize = 8, fontface = "bold"),
      labels_gp = gpar(fontsize = 8)
    ),
    Methylation = list(
      title_gp = gpar(fontsize = 8, fontface = "bold"),
      labels_gp = gpar(fontsize = 8)
    )))

heatmap_legend_param = list(title="Alteration", at = c("1"), title_gp = gpar(fontsize = 8), labels_gp = gpar(fontsize = 8))

#create the oncoprint
oncoprint <- oncoPrint(
    genetic_variations_red,
    alter_fun = list(
      background = function(x, y, w, h) {
        grid.rect(x, y, w, h, gp = gpar(fill = "#F8F8F8", col = "white"))
      },
      "1" = function(x, y, w, h) grid.rect(x, y, w, h, gp = gpar(fill = "black", col = "white")), col = "white"),
    col = c("1" = "black"), na_col = "white",
    remove_empty_rows = FALSE,
    remove_empty_columns = FALSE,
    row_names_gp = gpar(fontsize = 8),
    column_names_gp = gpar(fontsize = 8), 
    row_names_side = "left", pct_side = "right", pct_digits = 1,
    right_annotation = rowAnnotation(row_barplot = anno_oncoprint_barplot(c("1"), border = FALSE, height = unit(4, "cm"), axis_param = list(side = "bottom", labels_rot = 0))), 
    pct_gp = gpar(fontsize = 8), top_annotation = ha, heatmap_legend_param = heatmap_legend_param)
    
draw(oncoprint, heatmap_legend_side = "left", annotation_legend_side = "right", show_heatmap_legend = FALSE)

Fig2e <- grid.grabExpr(draw(oncoprint, heatmap_legend_side = "left", annotation_legend_side = "right", show_heatmap_legend = FALSE))

ggsave(Fig2e, path=(paste0(filepath,"figures")), filename = "Fig2e_CLL_genetic_landscape.svg")

```

# Drug response
## Comparison with previous data sets (Fig. 2B)
```{r, fig.width = 12, fig.height = 4}

#prepare data
screen <- screen |> 
  as.data.frame()

colnames(screen) <- colnames(screen) |> 
  gsub("\\.", "-", x = _) |>
  gsub("X10058-F4", "10058-F4", x = _) |> 
  gsub("D4476", "D 4476", x = _) |>
  gsub("IRAK1-4-Inhibitor-I", "IRAK-1/4 Inhibitor I", x = _) |> 
  gsub("Menin-MLL-Inhibitor", "Menin-MLL Inhibitor", x = _) |> 
  gsub("SB-203580", "SB 203580", x = _) |> 
  gsub("SGC-0946", "SGC 0946", x = _) |> 
  gsub("ZM-336372", "ZM 336372", x = _)

screen_long <- patients |> 
  as.data.frame() |> 
  arrange(patientID) |> 
  dplyr::select (patientID, diagnosis)  |> 
  merge(screen) |>  
  pivot_longer(cols = c(4:318), names_to = "drug", values_to = "viability")

#merge with concentrations by drug group (see Table S1)
concentrations_A <- c(10, 2, 0.4, 0.08, 0.016)
concentrations_B <- c(2, 0.4, 0.08, 0.016, 0.0032)
concentrations_C <- c(0.4, 0.08, 0.016, 0.0032, 0.00064)

drug_list <- list(
  A = concentrations_A,
  B = concentrations_B,
  C = concentrations_C
)

#add concentrations
screen_long <- screen_long |> 
  mutate(
    conc_index = as.numeric(str_extract(drug, "\\d+$")),
    drug = str_replace(drug, "_\\d+$", "")
    ) |> 
  left_join(drugs |> 
              dplyr::select(Drug, Group), by = c("drug" = "Drug")) |> 
  rowwise() |> 
  mutate(concentration = drug_list[[Group]][conc_index]) |> 
  ungroup()

#load data from IC50 screen (Dietrich et al. 2018)
IC50 <- read.csv("~/Dropbox/Medizin/Forschung/Aktuelle Forschungsprojekte/CLL CPS1000/raw data/IC50_170823.csv") #adjust

#preparate data for IC50 screen, create bins         
IC50 <- IC50 |>   
mutate(bin = cut(Concentration, breaks=c(0.004, 0.033, 0.156, 0.625, 2.5, 10, 40))) |> 
  filter(!is.na(bin))

#plot
density_cps1000 <- ggplot(screen_long, aes(viability)) + 
  geom_density(aes(fill=factor(concentration)), alpha=0.3, linewidth=0.25) + 
  labs(y="Densitiy", x="Viability", title="CPS1000", fill = expression(bold(paste("Concentration (", mu, "M)"))))+
  theme_figures+
  scale_fill_brewer(palette = "Blues", direction = 1)+
  scale_y_continuous(expand = c(0.02,0))+
  scale_x_continuous(limits=c(0,1.25), expand = c(0.02,0))

density_ic50 <- ggplot(IC50, aes(normVal)) + 
  geom_density(aes(fill=factor(bin)), alpha=0.3, linewidth=0.25) + 
  labs(y="Densitiy", x="Viability", title="IC50 (Dietrich et al. 2018)", fill = expression(bold(paste("Concentration (", mu, "M)"))))+
  theme_figures+
  scale_fill_brewer(palette = "Reds", direction = 1, labels = c("0.004 to 0.033", "0.033 to 0.156", "0.156 to 0.625", "0.625 to 2.5", "2.5 to 10", "10 to 40"))+
  scale_y_continuous(expand = c(0.02,0))+
  scale_x_continuous(limits=c(0,1.25), expand = c(0.02,0))

Fig2b <- density_cps1000 + density_ic50 & theme(plot.margin = margin(r=1, l=0.5, t=0.5, b=0.5, unit="cm"))
Fig2b

ggsave(path=(paste0(filepath,"figures")), filename = "Fig2c_density_plots.svg")

```

```{r, echo = FALSE}
#Fig. 2 multipanel
line1 <- plot_grid(Fig2a,Fig2c, ncol=2, align = "h", rel_widths = c(1.25,1), labels=c("A", "C"), label_size=14)
line2 <- plot_grid(Fig2b,Fig2d, ncol=2, align = "v", axis = "tb", labels=c("B","D"), label_size=14, rel_widths = c(6, 1))
line3 <- plot_grid(NULL, ncol=1, labels=c("E"), label_size=14)
line4 <- plot_grid(Fig2e, ncol=1)

Fig2 <- plot_grid(line1,line2, line3, line4, NULL, ncol=1, rel_heights = c(0.18,0.15,0.01,0.30,0.31)) & theme(plot.margin=margin(r=1,l=1,t=1,b=1,unit="cm"))

ggsave(path=(paste0(filepath,"figures")), filename = "Fig2.svg", width=42, height=59.4, units = "cm")
ggsave(path=(paste0(filepath,"figures")), filename = "Fig2.pdf", width=42, height=59.4, units = "cm")
ggsave(path=(paste0(filepath,"figures")), filename = "Fig2.png", width=42, height=59.4, units = "cm", dpi=600)

```

## Drug response patterns (Fig. 3A)
```{r, fig.width = 6, fig.height = 6}

#cap supravital viability results (cut-off <1.2)
screen_long$viability[screen_long$viability>1.2] <- 1.2
screen_long$viability[screen_long$viability<0] <- 0

#show diseases separately with at least 10 samples
pat_nr_per_disease <- patients |> 
  group_by(diagnosis) |> 
  summarize(n_pat = n()) |> 
  arrange(desc(n_pat))

diagnoses_show <- pat_nr_per_disease |> 
  filter(n_pat >=10) |> 
  pull(diagnosis)

df_tsne <- patients[,c("patientID", "diagnosis")] |> 
merge(screen, by="patientID") |> 
  dplyr::select(-patientID, -sampleID) |> 
  mutate(diagnosis= case_when(diagnosis %in% diagnoses_show ~ diagnosis,
                               TRUE ~ "other"))

#split dataframe in measures and subsetting
df_meas <- df_tsne[, !names(df_tsne) %in% "diagnosis"]
df_subset <- df_tsne[, names(df_tsne) %in% "diagnosis"]

#define order
order <- c("AML", "B-ALL", "B-PLL", "CLL", "MCL", "T-ALL", "T-PLL", "other")
df_subset <- factor(df_tsne$diagnosis, levels = order)

#t-sne
set.seed(12)
tsne_results <- Rtsne(df_meas, perplexity=25, check_duplicates = FALSE)

#plot
tsne_plot <- data.frame(
  x = tsne_results$Y[,1], 
  y = tsne_results$Y[,2],
  diagnosis = df_subset
)

Fig3a <- ggplot(tsne_plot) + 
  geom_point(aes(x=x, y=y, color=diagnosis), size=2, alpha=0.8) +
  theme_figures+
  labs(color="Diagnosis")+
  scale_color_manual(values=colors_diag)+
  theme(plot.margin = margin(t=2, b=2, r=1, l=1, unit="cm"))
Fig3a

ggsave(path=(paste0(filepath,"figures")), filename = "Fig3a_tsne.svg")

```

## Mean viability by drug and disease 
### Overall Heatmap (Fig. 3B)
```{r, fig.width = 12, fig.height = 18}

#prepare data
screen_long_topdiag <- screen_long |> 
  mutate(diagnosis = case_when(diagnosis %in% diagnoses_show ~ diagnosis, 
                               TRUE ~ "other"))

#calculate mean viability
screen_means <- screen_long_topdiag |> 
  group_by(patientID, diagnosis, drug) |> 
    summarize(mean_viab = mean(viability))
screen_means

#merge with metadata for IGHV status in CLL, drop CLL without IGHV annotation
ighv <- metadata[, names(metadata) %in% c("Patient.ID", "IGHV.status")]

screen_means_ighv <- merge(screen_means, ighv, by.x = "patientID", by.y= "Patient.ID") |>     
  mutate(diagnosis = case_when(IGHV.status == "U" ~ "U-CLL",
                        IGHV.status == "M" ~ "M-CLL",
                        TRUE ~ diagnosis)) |> 
  filter (diagnosis != "CLL") |> 
  dplyr::select(-IGHV.status) 

#prepare for heatmap
df <- screen_means_ighv |> 
  dplyr::select(-diagnosis) |> 
  pivot_wider(names_from = patientID, values_from = mean_viab)

df_use <- df |> 
  dplyr::select(-drug) |> 
  colnames()

df_subset <- df |> 
  dplyr::select(-drug)

#diagnoses
df_diagnoses <- screen_means_ighv |> 
  filter(drug == "10058-F4") |> 
  dplyr::select(patientID, diagnosis)

#add pathways
df_pathway <- drugs[, names(drugs) %in% c("Drug", "Pathway_mod")]

df_combined <- merge(df_pathway, df, , by.x = "Drug", by.y = "drug")

#order pathways and diagnoses
df_combined$Pathway_mod <- factor(df_combined$Pathway_mod,
    levels = c("AKT/mTOR", "B-cell receptor", "Bromodomain/PLK", "Chemotherapy", 
               "DNA damage response", "JAK/STAT", "MAPK", "PI3K", "Stress pathway", "other"))
               
df_diagnoses$diagnosis <- factor(df_diagnoses$diagnosis,
    levels = c("AML", "B-ALL", "B-PLL", "M-CLL", "MCL", "T-ALL", "T-PLL", "U-CLL", "other"))

#scale
x <- data.frame(df_subset)
y <- data.frame(t(scale(t(x))))

#create matrix
rownames(y) = df$drug
colnames(y) = colnames(df_subset)
z <- as.matrix(y)

#annotations
color_fun_reversed <- colorRamp2(c(4, median(z), -4), c("red", "white", "blue"))

ha = HeatmapAnnotation(
  Diagnosis = df_diagnoses$diagnosis, 
  annotation_name_gp = gpar(fontsize = 8), 
  col = list(Diagnosis = colors_diag_IGHV),
  annotation_legend_param = list(
    Diagnosis = list(
      title_gp = gpar(fontsize = 8, fontface = "bold"),
      labels_gp = gpar(fontsize = 8),
      nrow = 2
    )
  )
)

pway <- as.data.frame(df_combined[,1:2])
pway_ordered = pway[match(rownames(z), pway$Drug), ]

ha2 = rowAnnotation(
  Pathway = pway_ordered$Pathway_mod, 
  annotation_name_gp = gpar(fontsize = 8), 
  col = list(Pathway = colors_pathways_mod),
  annotation_legend_param = list(
    Pathway = list(
      title_gp = gpar(fontsize = 8, fontface = "bold"),
      labels_gp = gpar(fontsize = 8),
      nrow = 2
    )
  )
)

#heatmap
h1 <- Heatmap(z, name = "Z-score", show_row_names = TRUE, show_column_names=FALSE, row_names_gp = gpar(fontsize = 8), width = ncol(z)*unit(0.25, "mm"), height = nrow(z)*unit(3.5, "mm"), col = color_fun_reversed, top_annotation = ha, left_annotation = ha2, show_column_dend = FALSE, row_km = 4, row_title = NULL, 
show_parent_dend_line = FALSE, heatmap_legend_param = list(direction = "horizontal", title_gp = gpar(fontsize = 8, fontface = "bold"), labels_gp = gpar(fontsize = 8)))

set.seed(1234)

#set spacing to legend
ht_opt(HEATMAP_LEGEND_PADDING = unit(1.5, "cm"))

Fig3b <- grid.grabExpr(draw(h1, merge_legends = TRUE, heatmap_legend_side = "bottom"))

# Reset options
ht_opt(RESET = TRUE)  

#save?

```
### Drug-specific heatmaps (Fig. S2)

### Overview of drug effects (Fig. 3C)
```{r, fig.width = 12, fig.height = 4}

#create diagnosis-specific means
screen_mean_diag <- screen_long |> 
  group_by(diagnosis, drug) |> 
  summarize(mean_diag = mean(viability)) |> 
  filter(diagnosis %in% diagnoses_show) |> 
  pivot_wider(names_from = diagnosis, values_from = mean_diag)

#create heatmap
w <- screen_mean_diag[, names(screen_mean_diag) != "drug"]
rownames(w) = screen_mean_diag$drug
w <- as.matrix(t(w))

set.seed(123)
ha = rowAnnotation(foo = anno_text(rownames(w), gp = gpar(fontsize=8), just="right", offset=1))
column_ha = HeatmapAnnotation(foo = anno_text(colnames(w), gp = gpar(fontsize=8), just = "left", offset =0))

h2 <- Heatmap(w, name = "Mean viability", left_annotation = ha, 
top_annotation = column_ha, 
show_row_names = FALSE, show_column_names=FALSE, width = ncol(z)*unit(0.45, "mm"), height = nrow(z)*unit(0.4, "mm"), show_row_dend=TRUE, row_dend_side = "left", show_column_dend=FALSE,
row_dend_width = unit(7, "mm"), heatmap_legend_param = list(title_gp = gpar(fontsize = 8, fontface = "bold"), labels_gp = gpar(fontsize = 8)))

Fig3c_top <- draw(h2, heatmap_legend_side = "right", padding = unit(c(2, 2, 0, 2), "mm"))

#extract heatmap order
clustered_col_indices <- column_order(h2)
clustered_col_names <- colnames(w)[clustered_col_indices]

#create patient-specific means
screen_mean_pat <- screen_long |> 
  filter(diagnosis %in% diagnoses_show) |> 
  group_by(drug, patientID) |> 
  summarize(mean_pat = mean(viability)) |> 
  merge(drugs[,c("Drug", "Category")], by.x="drug", by.y="Drug") |> 
  mutate(drug = factor(drug, levels = clustered_col_names)) |> 
  arrange(drug)

#add boxplots  
Fig3c_bottom <- ggplot(screen_mean_pat, aes(x=drug, y=mean_pat, color=Category))+ 
  geom_boxplot(alpha=0.6, outlier.shape=NA) +
  geom_jitter(alpha=0.1, size=0.2, width = 0.15)+
  theme_figures+
  theme(axis.text.x = element_text(color="black", size=8, angle=90, hjust=1, vjust=0.5), 
        legend.position = "bottom",
        panel.spacing = unit(0, "mm"),
        plot.margin = margin(t=0, b=2, r=33, l=21, unit = "mm"))+
  scale_y_continuous(name ="Mean viability", breaks=seq(0,1.2, 0.2), limits=c(0,1.2)) +
  scale_x_discrete(name="")+
  scale_color_manual(values=colors_drug_type, name =  "Drug type", labels=c("Chemotherapy", "Kinase inhibitor", "Other"))
Fig3c_bottom

Fig3c_bottom_aligned <- ggdraw(Fig3c_bottom) +
  theme(plot.margin = margin(0, 0, 0, 0))

Fig3c <- plot_grid(grid.grabExpr(draw(h2, heatmap_legend_side = "right")), Fig3c_bottom_aligned, ncol=1, rel_heights = c(0.5, 0.5))

#save?
```

### Disease-specific drug effects 
#### Overview (Fig. S3 and S4)
```{r}

#create plots
aml <- create_boxplot(screen_means_ighv, "AML", colors_diag_IGHV, "AML")
ball <- create_boxplot(screen_means_ighv, "B-ALL", colors_diag_IGHV, "B-ALL")
tall <- create_boxplot(screen_means_ighv, "T-ALL", colors_diag_IGHV, "T-ALL")

SFig3 <- plot_grid(aml, ball, tall, NULL, NULL, ncol = 5, nrow = 1)
ggsave(path=(paste0(filepath,"figures")), filename = "SFig3_viab_leucemia.svg", width=42, height=20, units = "cm")


bpll <- create_boxplot(screen_means_ighv, "B-PLL", colors_diag_IGHV, "B-PLL")
mcll <- create_boxplot(screen_means_ighv, "M-CLL", colors_diag_IGHV, "M-CLL")
mcl <- create_boxplot(screen_means_ighv, "MCL", colors_diag_IGHV, "MCL")
ucll <- create_boxplot(screen_means_ighv, "U-CLL", colors_diag_IGHV, "U-CLL")
tpll <- create_boxplot(screen_means_ighv, "T-PLL", colors_diag_IGHV, "T-PLL")

SFig4 <- plot_grid(bpll, mcll, mcl, ucll, tpll, ncol = 5, nrow = 1)
ggsave(path=(paste0(filepath,"figures")), filename = "SFig4_viab_lymphoma.svg", width=42, height=20, units = "cm")

```

#### Heatmap (Fig. 3D)
```{r, fig.width = 8, fig.height = 8}

#prepare data
diagnoses_show_ighv <- c(diagnoses_show, "U-CLL", "M-CLL")
diagnoses_show_ighv <- diagnoses_show_ighv[diagnoses_show_ighv != "CLL"]

viabTab.mean <- screen_means_ighv |> 
  filter(diagnosis %in% diagnoses_show_ighv)

#create p-values
allDiag <- unique(viabTab.mean$diagnosis)

pTabAll <- lapply(allDiag, function(diagBase) {
  #calculate p value matrix
  pTab <- lapply(allDiag[allDiag != diagBase], function(diagCompare) {
    testTab <- filter(viabTab.mean, diagnosis %in% c(diagBase, diagCompare))  |> 
      mutate(diagnosis = factor(diagnosis, levels = c(diagBase,diagCompare))) 
    res <- group_by(testTab, drug) |> do(tTest(.$mean_viab, .$diagnosis)) |>
      mutate(diagCompare = diagCompare, diagBase = diagBase)
    res
  }) |> bind_rows() |> ungroup() |> mutate(p.adj = p.adjust(p, method = "BH"))
}) |> bind_rows()

#prepare data table for plot 
plotTab <- pTabAll |> mutate(p = ifelse(p < 1e-12, 1e-12, p)) |>
  mutate(pSign = -log10(p)*sign(diff)) |>
  mutate(pSign = ifelse(p.adj < 0.1, pSign, 0)) |>
  ungroup()

#order drugs by their association similarity (hierarchical clustering)
pMat <- mutate(plotTab, diagPair = paste0(diagBase,"_",diagCompare)) |>
  dplyr::select(drug, pSign, diagPair) |>
  spread(key = diagPair, value = pSign) |> data.frame() |>
  column_to_rownames("drug") |> as.matrix()

#cut
set.seed(123)
nClust = 9
hcRes <- hclust(dist(pMat), method = "ward.D2")
drugOrder <- rownames(pMat)[hcRes$order]
clustMap <- cutree(hcRes,nClust)

diagOrder <- c("U-CLL", "M-CLL","MCL", "B-PLL", "T-PLL", "B-ALL", "T-ALL","AML")
plotTab <- mutate(plotTab, drug = factor(drug, levels = drugOrder), 
                  diagBase = factor(diagBase, levels= diagOrder), 
                  diagCompare = factor(diagCompare, levels = diagOrder)) |>
  arrange(drug) |> mutate(drugClust = paste0("",nClust-clustMap[as.character(drug)]+1))

#color strips in y-direction
strip <- strip_themed(background_y = elem_list_rect(fill = "white"))

#add pathway annotation
plotTab <- merge(plotTab, drugs[,c("Drug", "Pathway_mod")], by.x="drug", by.y="Drug")

#plot
main_plot <- ggplot(plotTab, aes(x=diagCompare, y = drug, fill = pSign)) + 
  geom_tile(size = 0.3, color = "white") +
  facet_grid2(rows = vars(drugClust), cols = vars(diagBase), scales = "free", space = "free_y", strip = strip) + 
  scale_fill_gradient2(
    high = "blue", mid = "white", low = "red", midpoint = 0)+
  theme_figures_facet_angle60_x+
  labs(fill = "**-log<sub>10</sub>*P*<br>with direction**") +
  theme(legend.title = ggtext::element_markdown())+
  ylab("") + xlab("")
main_plot

#add pathway annotation
anno_plot <- ggplot(plotTab, aes(x = diagCompare, y = drug, fill = Pathway_mod)) +
  geom_tile(size = 0.3) +
  facet_grid(rows = vars(drugClust), scales = "free", space = "free_y") +
  scale_fill_manual(values = colors_pathways_mod,
                    name = "Drug Class")+
  theme_minimal() +
  theme(axis.text = element_blank(),
        axis.title = element_blank(),
        panel.grid = element_blank(),
        panel.border = element_rect(linewidth = 0.75),
        panel.background = element_blank(),
        plot.background = element_blank(),
        strip.text.y = element_blank(),
        panel.spacing = unit(0.15, "lines"), 
        legend.position = "none")

Fig3d <- ggdraw() + 
  draw_plot(main_plot, x=0,y=0,width=1,height=1)+
  draw_plot(anno_plot,x=0.8612,y=0.0765,width=0.03,height=0.895) & 
  theme(plot.margin = margin(t=1, unit="cm"))

ggsave(path=(paste0(filepath,"figures")), filename = "Fig3d_pvalue_heatmap.svg", width=42, height=30, units = "cm")

```

#### Drug ranking
```{r}

#prepare data
#rank drugs
rankTab <- screen_long |> 
  group_by(patientID, drug, diagnosis) |> 
  summarise(meanViab = mean(viability)) |> 
  dplyr::ungroup()
medTab <- screen_long |> 
  group_by(drug) |> 
  summarise(medVal = median(viability))
rankTab <- left_join(rankTab, medTab, by = "drug") |>  
  mutate(score = meanViab - medVal)

#make rank plot
rank_score_tab <- rankTab |> 
  group_by(patientID, diagnosis) |> 
  arrange((score)) |> 
  mutate(rank_score=row_number(), l2fc = log2(meanViab)-log2(medVal)) |> 
  ungroup()

#show ranks for AML with venetoclax and cytarabine
aml_cyt <- rank_score_tab |> 
filter(diagnosis %in% c("AML")) |> 
mutate(drug_col = ifelse(drug %in% c("Cytarabine"), drug, "other")) |>
ggplot(aes(x=rank_score, y=score, alpha=drug_col, group=patientID))+
geom_hline(yintercept = 0, color="grey70")+
geom_line(alpha=0.25, color="black")+
geom_point(aes(alpha=drug_col, fill=drug_col, color=drug_col), size=1.5, shape=21)+
scale_alpha_manual(values=c(0.75,0))+
scale_fill_manual(name="Drug", breaks=c("Cytarabine"), values=c("#E31A1C", ""))+
scale_color_manual(name="Drug", breaks=c("Cytarabine"), values=c("#E31A1C", ""))+
  labs(title="Cytarabine", y="Score", x="Drug rank")+
  theme_bw()+
  theme(axis.text = element_text(color="black", size=8),
  legend.position = "none", axis.title = element_text(size=8), plot.title = element_text(hjust=0.5, size=8),
  panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    panel.border = element_rect(color = "black", linewidth = 1, fill = NA))

aml_ven <- rank_score_tab |> 
filter(diagnosis %in% c("AML")) |> 
mutate(drug_col = ifelse(drug %in% c("Venetoclax"), drug, "other")) |>
ggplot(aes(x=rank_score, y=score, alpha=drug_col, group=patientID))+
geom_hline(yintercept = 0, color="grey70")+
geom_line(alpha=0.25, color="black")+
geom_point(aes(alpha=drug_col, fill=drug_col, color=drug_col), size=1.5, shape=21)+
scale_alpha_manual(values=c(0, 0.75))+
scale_fill_manual(name="Drug", breaks=c("Venetoclax"), values=c("#E31A1C", ""))+
scale_color_manual(name="Drug", breaks=c("Venetoclax"), values=c("#E31A1C", ""))+
  labs(title="Venetoclax", y="Score", x="Drug rank")+
  theme_bw()+
  theme(axis.text = element_text(color="black", size=8),
  legend.position = "none", axis.title = element_text(size=8), plot.title = element_text(hjust=0.5, size=8),
  panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    panel.border = element_rect(color = "black", linewidth = 1, fill = NA))

SFig8c <- aml_cyt + aml_ven + plot_annotation(title = "AML", theme = theme(plot.title = element_text(hjust=0.5, size=8, face="bold"))) + theme(plot.margin = margin(t=1, unit = "cm"))
SFig8c

#combine with IGHV annotation
cll_meta <- metadata[, c("Patient.ID", "diagnosis", "IGHV.status")] |> 
  filter(diagnosis == "CLL")

rank_score_tab_mod <- merge(rank_score_tab, cll_meta[,c("Patient.ID", "IGHV.status")], by.x="patientID", by.y="Patient.ID", all.x = TRUE) |> 
  mutate(diagnosis = case_when(IGHV.status == "U" ~ "U-CLL",
                               IGHV.status == "M" ~ "M-CLL",
                               TRUE ~ diagnosis)) |> 
                               filter(diagnosis != "CLL")

drug_ranking_diag <- rank_score_tab_mod |> 
  group_by(drug, diagnosis) |> 
  summarize(median_score = median(score), score_sd = sd(score)) |> 
  arrange(median_score) |> 
  group_by(diagnosis) |>
  mutate(rank_median = row_number())
  
# plot: aml
drug_order_aml <- drug_ranking_diag |> 
  filter(diagnosis == "AML") |> 
  arrange(median_score) |> 
  pull(drug)
  
drug_ranking_diag$drug <- factor(drug_ranking_diag$drug, levels=drug_order_aml)  

rank_aml <- drug_ranking_diag |> 
  filter(diagnosis == "AML") |> 
  mutate(direction = sign(median_score)) |> 
  ggplot(aes(x=drug, y=median_score, fill=diagnosis))+
  geom_col(alpha=0.9) +
  geom_errorbar(aes(ymin = median_score-score_sd, ymax = median_score+score_sd), 
  color="black", alpha=0.75, width=0.5, linewidth=0.5)+
  facet_wrap(~diagnosis) +
  labs(title="", y="Median score", x="") +
  theme_bw() +
  guides(color="none")+
  scale_fill_manual(values=colors_diag)+
    theme(legend.position = "none", 
          axis.text.x = element_text(angle = 90, hjust=1, vjust=0.5, color="black", size=8), 
          panel.grid = element_line(colour = "white"), 
          axis.text.y = element_text(color="black", size=8), 
          axis.title.y = element_text(size=8), 
          axis.title.x = element_text(size=8), 
          panel.border = element_rect(color = "black", linewidth = 1, fill = NA),
          strip.background = element_rect(color = "black", linewidth = 1, fill = "grey90"),
          plot.title = element_text(hjust=0.5, size=8, face="bold"), 
          strip.text = element_text(size = 8), 
          plot.margin = margin(l = 0.5, r = 1, unit = "cm"))
          
# plot: u-cll
drug_order_ucll <- drug_ranking_diag |> 
  filter(diagnosis == "U-CLL") |> 
  arrange(median_score) |> 
  pull(drug)
  
drug_ranking_diag$drug <- factor(drug_ranking_diag$drug, levels=drug_order_ucll)  

rank_ucll <- drug_ranking_diag |> 
  filter(diagnosis == "U-CLL") |> 
  mutate(direction = sign(median_score)) |> 
  ggplot(aes(x=drug, y=median_score, fill=diagnosis))+
  geom_col(alpha=0.9) +
  geom_errorbar(aes(ymin = median_score-score_sd, ymax = median_score+score_sd), 
  color="black", alpha=0.75, width=0.5, linewidth=0.5)+
  facet_wrap(~diagnosis) +
  labs(title="", y="", x="") +
  theme_bw() +
  guides(color="none")+
  scale_fill_manual(values=colors_diag_IGHV)+
    theme(legend.position = "none", 
          axis.text.x = element_text(angle = 90, hjust=1, vjust=0.5, color="black", size=8), 
          panel.grid = element_line(colour = "white"), 
          axis.text.y = element_text(color="black", size=8), 
          axis.title.y = element_text(size=8), 
          axis.title.x = element_text(size=8), 
          panel.border = element_rect(color = "black", linewidth = 1, fill = NA),
          strip.background = element_rect(color = "black", linewidth = 1, fill = "grey90"),
          plot.title = element_text(hjust=0.5, size=8, face="bold"), 
          strip.text = element_text(size = 8), 
          plot.margin = margin(l = 0.5, r = 1, unit = "cm"))    
          
# plot: m-cll
drug_order_mcll <- drug_ranking_diag |> 
  filter(diagnosis == "M-CLL") |> 
  arrange(median_score) |> 
  pull(drug)
  
drug_ranking_diag$drug <- factor(drug_ranking_diag$drug, levels=drug_order_mcll)  

rank_mcll <- drug_ranking_diag |> 
  filter(diagnosis == "M-CLL") |> 
  mutate(direction = sign(median_score)) |> 
  ggplot(aes(x=drug, y=median_score, fill=diagnosis))+
  geom_col(alpha=0.9) +
  geom_errorbar(aes(ymin = median_score-score_sd, ymax = median_score+score_sd), 
  color="black", alpha=0.75, width=0.5, linewidth=0.5)+
  facet_wrap(~diagnosis) +
  labs(title="", y="", x="") +
  theme_bw() +
  guides(color="none")+
  scale_fill_manual(values=colors_diag_IGHV)+
    theme(legend.position = "none", 
          axis.text.x = element_text(angle = 90, hjust=1, vjust=0.5, color="black", size=8), 
          panel.grid = element_line(colour = "white"), 
          axis.text.y = element_text(color="black", size=8), 
          axis.title.y = element_text(size=8), 
          axis.title.x = element_text(size=8), 
          panel.border = element_rect(color = "black", linewidth = 1, fill = NA),
          strip.background = element_rect(color = "black", linewidth = 1, fill = "grey90"),
          plot.title = element_text(hjust=0.5, size=8, face="bold"), 
          strip.text = element_text(size = 8), 
          plot.margin = margin(l = 0.5, r = 1, unit = "cm"))            
          
# plot: b-all
drug_order_ball <- drug_ranking_diag |> 
  filter(diagnosis == "B-ALL") |> 
  arrange(median_score) |> 
  pull(drug)
  
drug_ranking_diag$drug <- factor(drug_ranking_diag$drug, levels=drug_order_ball)  

rank_ball <- drug_ranking_diag |> 
  filter(diagnosis == "B-ALL") |> 
  mutate(direction = sign(median_score)) |> 
  ggplot(aes(x=drug, y=median_score, fill=diagnosis))+
  geom_col(alpha=0.9) +
  geom_errorbar(aes(ymin = median_score-score_sd, ymax = median_score+score_sd), 
  color="black", alpha=0.75, width=0.5, linewidth=0.5)+
  facet_wrap(~diagnosis) +
  labs(title="", y="", x="") +
  theme_bw() +
  guides(color="none")+
  scale_fill_manual(values=colors_diag)+
    theme(legend.position = "none", 
          axis.text.x = element_text(angle = 90, hjust=1, vjust=0.5, color="black", size=8), 
          panel.grid = element_line(colour = "white"), 
          axis.text.y = element_text(color="black", size=8), 
          axis.title.y = element_text(size=8), 
          axis.title.x = element_text(size=8), 
          panel.border = element_rect(color = "black", linewidth = 1, fill = NA),
          strip.background = element_rect(color = "black", linewidth = 1, fill = "grey90"),
          plot.title = element_text(hjust=0.5, size=8, face="bold"), 
          strip.text = element_text(size = 8), 
          plot.margin = margin(l = 0.5, r = 1, unit = "cm"))      
          
# plot: t-all
drug_order_tall <- drug_ranking_diag |> 
  filter(diagnosis == "T-ALL") |> 
  arrange(median_score) |> 
  pull(drug)
  
drug_ranking_diag$drug <- factor(drug_ranking_diag$drug, levels=drug_order_tall)  

rank_tall <- drug_ranking_diag |> 
  filter(diagnosis == "T-ALL") |> 
  mutate(direction = sign(median_score)) |> 
  ggplot(aes(x=drug, y=median_score, fill=diagnosis))+
  geom_col(alpha=0.9) +
  geom_errorbar(aes(ymin = median_score-score_sd, ymax = median_score+score_sd), 
  color="black", alpha=0.75, width=0.5, linewidth=0.5)+
  facet_wrap(~diagnosis) +
  labs(title="", y="Median score", x="") +
  theme_bw() +
  guides(color="none")+
  scale_fill_manual(values=colors_diag)+
    theme(legend.position = "none", 
          axis.text.x = element_text(angle = 90, hjust=1, vjust=0.5, color="black", size=8), 
          panel.grid = element_line(colour = "white"), 
          axis.text.y = element_text(color="black", size=8), 
          axis.title.y = element_text(size=8), 
          axis.title.x = element_text(size=8), 
          panel.border = element_rect(color = "black", linewidth = 1, fill = NA),
          strip.background = element_rect(color = "black", linewidth = 1, fill = "grey90"),
          plot.title = element_text(hjust=0.5, size=8, face="bold"), 
          strip.text = element_text(size = 8), 
          plot.margin = margin(l = 0.5, r = 1, unit = "cm"))            

# plot: mcl
drug_order_mcl <- drug_ranking_diag |> 
  filter(diagnosis == "MCL") |> 
  arrange(median_score) |> 
  pull(drug)
  
drug_ranking_diag$drug <- factor(drug_ranking_diag$drug, levels=drug_order_mcl)  

rank_mcl <- drug_ranking_diag |> 
  filter(diagnosis == "MCL") |> 
  mutate(direction = sign(median_score)) |> 
  ggplot(aes(x=drug, y=median_score, fill=diagnosis))+
  geom_col(alpha=0.9) +
  geom_errorbar(aes(ymin = median_score-score_sd, ymax = median_score+score_sd), 
  color="black", alpha=0.75, width=0.5, linewidth=0.5)+
  facet_wrap(~diagnosis) +
  labs(title="", y="Median score", x="") +
  theme_bw() +
  guides(color="none")+
  scale_fill_manual(values=colors_diag)+
    theme(legend.position = "none", 
          axis.text.x = element_text(angle = 90, hjust=1, vjust=0.5, color="black", size=8), 
          panel.grid = element_line(colour = "white"), 
          axis.text.y = element_text(color="black", size=8), 
          axis.title.y = element_text(size=8), 
          axis.title.x = element_text(size=8), 
          panel.border = element_rect(color = "black", linewidth = 1, fill = NA),
          strip.background = element_rect(color = "black", linewidth = 1, fill = "grey90"),
          plot.title = element_text(hjust=0.5, size=8, face="bold"), 
          strip.text = element_text(size = 8), 
          plot.margin = margin(l = 0.5, r = 1, unit = "cm"))  

# plot: tpll
drug_order_tpll <- drug_ranking_diag |> 
  filter(diagnosis == "T-PLL") |> 
  arrange(median_score) |> 
  pull(drug)
  
drug_ranking_diag$drug <- factor(drug_ranking_diag$drug, levels=drug_order_tpll) 

rank_tpll <- drug_ranking_diag |> 
  filter(diagnosis == "T-PLL") |> 
  mutate(direction = sign(median_score)) |> 
  ggplot(aes(x=drug, y=median_score, fill=diagnosis))+
  geom_col(alpha=0.9) +
  geom_errorbar(aes(ymin = median_score-score_sd, ymax = median_score+score_sd), 
  color="black", alpha=0.75, width=0.5, linewidth=0.5)+
  facet_wrap(~diagnosis) +
  labs(title="", y="", x="") +
  theme_bw() +
  guides(color="none")+
  scale_fill_manual(values=colors_diag)+
    theme(legend.position = "none", 
          axis.text.x = element_text(angle = 90, hjust=1, vjust=0.5, color="black", size=8), 
          panel.grid = element_line(colour = "white"), 
          axis.text.y = element_text(color="black", size=8), 
          axis.title.y = element_text(size=8), 
          axis.title.x = element_text(size=8), 
          panel.border = element_rect(color = "black", linewidth = 1, fill = NA),
          strip.background = element_rect(color = "black", linewidth = 1, fill = "grey90"),
          plot.title = element_text(hjust=0.5, size=8, face="bold"), 
          strip.text = element_text(size = 8), 
          plot.margin = margin(l = 0.5, r = 1, unit = "cm"))  
          

# plot: bpll
drug_order_bpll <- drug_ranking_diag |> 
  filter(diagnosis == "B-PLL") |> 
  arrange(median_score) |> 
  pull(drug)
  
drug_ranking_diag$drug <- factor(drug_ranking_diag$drug, levels=drug_order_bpll) 

rank_bpll <- drug_ranking_diag |> 
  filter(diagnosis == "B-PLL") |> 
  mutate(direction = sign(median_score)) |> 
  ggplot(aes(x=drug, y=median_score, fill=diagnosis))+
  geom_col(alpha=0.9) +
  geom_errorbar(aes(ymin = median_score-score_sd, ymax = median_score+score_sd), 
  color="black", alpha=0.75, width=0.5, linewidth=0.5)+
  facet_wrap(~diagnosis) +
  labs(title="", y="Median score", x="") +
  theme_bw() +
  guides(color="none")+
  scale_fill_manual(values=colors_diag)+
    theme(legend.position = "none", 
          axis.text.x = element_text(angle = 90, hjust=1, vjust=0.5, color="black", size=8), 
          panel.grid = element_line(colour = "white"), 
          axis.text.y = element_text(color="black", size=8), 
          axis.title.y = element_text(size=8), 
          axis.title.x = element_text(size=8), 
          panel.border = element_rect(color = "black", linewidth = 1, fill = NA),
          strip.background = element_rect(color = "black", linewidth = 1, fill = "grey90"),
          plot.title = element_text(hjust=0.5, size=8, face="bold"), 
          strip.text = element_text(size = 8), 
          plot.margin = margin(l = 0.5, r = 1, unit = "cm"))          

SFig8b <- (rank_aml+plot_spacer()+rank_ball+plot_layout(widths = c(1, -0.1 ,1)))/(rank_bpll+plot_spacer()+rank_mcll+plot_layout(widths = c(1, -0.1 ,1)))/(rank_mcl+plot_spacer()+rank_ucll+plot_layout(widths = c(1, -0.1 ,1)))/(rank_tall+plot_spacer()+rank_tpll+plot_layout(widths = c(1, -0.1 ,1))) + theme(plot.margin = margin(t=1, unit = "cm"))

#calculate global ranks
#show all diseases
global_rank_tab <- rank_score_tab_mod |> 
group_by(patientID) |> 
mutate(global_score = sum(score)) |> 
arrange(global_score) |> 
group_by(drug) |> 
mutate(global_rank = row_number()) |> 
filter(drug == "Ibrutinib") |> 
#calculate median global rank per diagnosis
group_by(diagnosis) |> 
mutate(global_score_median_diag = median(global_score),
        global_rank_median_diag = median(global_rank))
        

#plot
diagSelected <- c("AML", "B-ALL", "T-ALL", "T-PLL", "B-PLL", "MCL", "U-CLL", "M-CLL")

global_rank_tab <- global_rank_tab |> 
  mutate(diag_col = ifelse(diagnosis %in% diagSelected, diagnosis, "other"))
  
global_rank_tab$diag_col <- factor(global_rank_tab$diag_col, levels = sort(diagSelected))

# Calculate the line for all data
reference_line <- global_rank_tab |>
  arrange(global_rank) |>
  dplyr::select(global_rank, global_score)
  
global_rank_tab <- global_rank_tab |> 
  filter(!is.na(diag_col)) 

score_ranking <- global_rank_tab |> 
  arrange(global_score_median_diag) |> 
  distinct(diag_col) |> 
  pull(diag_col)
  
global_rank_tab$diag_col <- factor(global_rank_tab$diag_col, levels = score_ranking)

# Now plot with facets
all_diag <- global_rank_tab |> 
  ggplot(aes(x=global_rank, y=global_score)) +
  geom_hline(data = global_rank_tab |> 
               distinct(diag_col, global_score_median_diag),
             aes(yintercept = global_score_median_diag), 
             color="grey90")+
  geom_line(data = reference_line, color="grey30", linewidth=0.5, inherit.aes=TRUE) +
  geom_hline(data = global_rank_tab |> 
               distinct(diag_col, global_score_median_diag),
             aes(yintercept = global_score_median_diag), 
             color="grey80")+
  geom_point(aes(color=diag_col), alpha=0.8, size=1.5) +
  #geom_point(aes(x=global_rank_median_diag, y=global_score_median_diag), color="grey80",size=1.5, shape=21) +
  facet_wrap(~diag_col, ncol=4) +
  scale_alpha_manual(values=c(rep(0.5, length(diagSelected)))) +
  scale_color_manual(values=colors_diag_IGHV)+
  labs(title="", y="Sum of drug scores", x="Sample rank") +
  theme_bw() +
  guides(alpha = "none", color="none")+
    theme(legend.position = "none", 
          axis.text.x = element_text(angle = 0, color="black", size=8), 
          panel.grid = element_line(colour = "white"), 
          axis.text.y = element_text(color="black", size=8), 
          axis.title.y = element_text(size=8), 
          axis.title.x = element_text(size=8), 
          panel.border = element_rect(color = "black", linewidth = 1, fill = NA),
          strip.background = element_rect(color = "black", linewidth = 1, fill = "grey90"),
          plot.title = element_text(hjust=0.5, size=8, face="bold"), 
          strip.text = element_text(size = 8), 
          plot.margin = margin(l = 0.5, r = 1, unit = "cm"))

SFig8a <- all_diag + theme(plot.margin = margin(t=1, l=1, r=1, unit = "cm"))

#multipanel
row3 <- plot_grid(SFig8c, NULL, ncol = 2, rel_widths = c(0.4, 0.6))

SFig.5 <- plot_grid(SFig8a, SFig8b, row3, rel_heights = c(0.2, 0.65, 0.15), ncol = 1, 
                   align = "v", axis = c("l", "r"), 
                   labels=c("A", "B", "C"), label_size=14) & theme(plot.margin=margin(r=1,l=1,t=1,b=1,unit="cm"))


ggsave(path=(paste0(filepath,"figures")), filename = "FigS5.svg", width=42, height=59.4, units = "cm")
ggsave(path=(paste0(filepath,"figures")), filename = "FigS5.pdf", width=42, height=59.4, units = "cm")
ggsave(path=(paste0(filepath,"figures")), filename = "FigS5.png", width=42, height=59.4, units = "cm", dpi=600) #remove later

```



#### Signature drugs at the single concentration level (multivariate binary logistic regression)
```{r}



```



#### Specific drugs
```{r}

#prepare data
drugs_show <- c("Ibrutinib", "PF-3758309", "AZD8055", "Venetoclax", "Duvelisib") 
drug_targets <- c("Ibrutinib" = "BTK", "PF-3758309" = "PAK", "AZD8055" = "mTOR", "Venetoclax" = "BCL2", "Duvelisib" = "PI3K")

screen_long_selected_drugs <- screen_long |> 
  filter(drug %in% drugs_show) |> 
  group_by(drug, diagnosis, conc_index) |>
    mutate(mean_viab=mean(viability), 
    std.error=std.error(viability))

screen_long_selected_drugs$concentration <- as.character(screen_long_selected_drugs$concentration)

#show drug-response for BTK, PAK and mTOR inhibitor
#helper function for plots
viab_plot_list <- list()

viab_plot_list <- map(drugs_show, ~{
  screen_long_selected_drugs |> 
    filter(drug == .x, diagnosis %in% diagnoses_show) |> 
    ggplot(aes(x=concentration, y=mean_viab, group=diagnosis, color=diagnosis)) +
      geom_errorbar(aes(ymin=mean_viab-std.error, ymax=mean_viab+std.error), 
                    width=.5, position=position_dodge(0.01)) +
      geom_line() +
      theme_figures_angle45_x +
      labs(title=paste0(.x, " (", drug_targets[.x], ")"), 
           y="Viability", x=bquote("Concentration ("*mu*"M)"), 
           color="Diagnosis") +
      scale_y_continuous(breaks=seq(0,1.0, 0.2), limits=c(0,1.1)) +
      scale_color_manual(values=colors_diag)
}) |> 
  set_names(drugs_show)

#combine plots
Fig3e <- viab_plot_list$Ibrutinib + viab_plot_list$`PF-3758309` + viab_plot_list$AZD8055 + 
  plot_layout(ncol=3, guides = 'collect') & 
  theme(legend.position='right', legend.title= element_text(size=8, face="bold"), 
        plot.margin = margin(t=0.5, b=1, l=0.5, r=0.5, unit = "cm")) &
  guides(
    color = guide_legend(ncol = 1),
    fill = guide_legend(ncol = 1),
    shape = guide_legend(ncol = 1),
    linetype = guide_legend(ncol = 1)
  )
Fig3e

ggsave(path=(paste0(filepath,"figures")), filename = "Fig3e_viab_btk_pak_mtor.svg")
            
#show drug-response for BCL2 inhibtor (venetoclax)
diagnoses_ven <- c("CLL", "AML", "B-ALL", "T-ALL", "MCL", "Sezary")

screen_long_ven <- screen_long_selected_drugs |> 
  filter(drug == "Venetoclax", diagnosis %in% diagnoses_ven) |> 
  group_by(diagnosis, concentration) |> 
  summarize(med_viab=median(viability))

#fit dose-response models
screen_long_ven$concentration <- as.numeric(screen_long_ven$concentration)

models_ven <- map(diagnoses_ven, function(dx) {
  data <- screen_long_ven |> filter(diagnosis == dx)
  drm(med_viab ~ concentration, 
      data = data, 
      fct = LL.4(names = c("hill", "min_value", "max_value", "ec_50")))
})
names(models_ven) <- tolower(diagnoses_ven)

#create predictions based on models
predictions_ven <- map(diagnoses_ven, function(dx) {
  data <- screen_long_ven |> filter(diagnosis == dx)
  model <- models_ven[[tolower(dx)]]
  
  new_data <- data.frame(
    concentration = seq(min(data$concentration), max(data$concentration), length.out = 100)
  )
  new_data$pred <- predict(model, newdata = new_data)
  new_data$diagnosis <- dx
  new_data
}) |> 
  bind_rows()

#calculate IC50 values for all diagnoses
ic50_ven <- do.call(rbind, lapply(diagnoses_ven, function(dx) {
  model <- models_ven[[tolower(dx)]]
  coefs <- setNames(model$coefficients, c("hill", "min_value", "max_value", "ec_50"))
  
  ic_50 <- with(as.list(coefs), 
                exp(log(ec_50) + (1 / hill) * log(max_value / (max_value - 2 * min_value))))
  
  data.frame(ic50 = ic_50, diagnosis = dx)
}))

#set colors
colors_ven <- c(setNames("darkgrey", "Sezary"), colors_diag)

#order of diagnoses
order_ven <- ic50_ven |> 
  arrange(ic50) |> 
  pull(diagnosis)

#plot
data_ven <- screen_long_selected_drugs |> 
  filter(drug == "Venetoclax", diagnosis %in% diagnoses_ven)
data_ven$diagnosis <- factor(data_ven$diagnosis, levels = order_ven)
data_ven$concentration <- as.numeric(data_ven$concentration)

predictions_ven$diagnosis <- factor(predictions_ven$diagnosis, levels = order_ven)
ic50_ven$diagnosis        <- factor(ic50_ven$diagnosis, levels = order_ven)

Fig3f <- data_ven |> 
  ggplot(aes(x=concentration, y=viability, group=diagnosis, color=diagnosis)) +
    geom_boxplot(aes(group=concentration), alpha = 0.8, outlier.shape = NA) +
    geom_jitter(aes(group=concentration), alpha=0.2, size=1, width = 0.15) +
    facet_wrap(~diagnosis, ncol=6) +
    geom_line(data = predictions_ven, aes(x = concentration, y = pred, color = diagnosis))+
    geom_vline(data = ic50_ven, aes(xintercept = ic50), 
               linetype = "dashed", color = "black", size = 0.4) +
    scale_x_log10(breaks = c(0.001, 0.01, 0.1, 1),
                  labels = trans_format("log10", math_format(10^.x)))+
    scale_y_continuous(breaks=seq(0,1.0, 0.2), limits=c(0,1.1)) +
    theme_figures_facet+
    labs(title="Venetoclax (BCL2)", x=bquote("Concentration ("*mu*"M)"), y = "Viability") +
    scale_color_manual(values=colors_ven)+
    theme(plot.margin = margin(t=1, b=1, r=1, l=1, unit="cm"))
Fig3f

ggsave(path=(paste0(filepath,"figures")), filename = "Fig3f_venetoclax.svg")

#duvelisib
Fig3g <- screen_long_selected_drugs |> 
  filter(diagnosis %in% c("B-ALL", "T-ALL"), drug == "Duvelisib") |> 
  ggplot(aes(x=factor(concentration), y=viability, group=interaction(diagnosis, concentration), color=diagnosis)) +
  geom_boxplot(alpha = 0.8, outlier.shape = NA)+
  geom_jitter(alpha = 0.2, size=1, position = position_jitterdodge())+
  labs(title="Duvelisib (PI3K)", y="Viability", x=bquote("Concentration ("*mu*"M)"), color = "Diagnosis")+
  scale_y_continuous(breaks=seq(0,1.0, 0.2), limits=c(0,1.1))+
  theme_figures_angle45_x+
  scale_color_manual(values=colors_diag)+
  theme(plot.margin = margin(t=1, b=1, r=1, l=1, unit="cm"))
Fig3g 
  
ggsave(path=(paste0(filepath,"figures")), filename = "Fig3g_duvelisib.svg")

#cobimetinib (...)

#quizartinib (...)
```

```{r, echo = FALSE}

#Fig. 3 multipanel
line1 <- plot_grid(Fig3a, Fig3c , ncol=2, rel_widths = c(0.35, 0.65), labels=c("A", "C"), label_size=14)
figb <- plot_grid(NULL, Fig3b, NULL, ncol=1, rel_heights = c(0.05, 0.925, 0.025))
figde <- plot_grid(Fig3d, Fig3e, ncol=1, rel_heights = c(0.75, 0.25), labels=c("D", "E"), label_size=14)
line2 <- plot_grid(figb, figde, ncol=2, rel_widths = c(0.42, 0.58), labels=c("B", ""), label_size=14)
line3 <- plot_grid(Fig3f, Fig3g, rel_widths = c(2.5, 1), align = "h", axis= "b", labels=c("F", "G"), label_size=14)

Fig3 <- plot_grid(line1, line2, line3, NULL, nrow=4, rel_heights = c(0.25, 0.5, 0.15, 0.1)) & theme(plot.margin=margin(r=1,l=1,t=1,b=1,unit="cm"))

ggsave(path=(paste0(filepath,"figures")), filename = "Fig3.svg", width=42, height=59.4, units = "cm")
ggsave(path=(paste0(filepath,"figures")), filename = "Fig3.pdf", width=42, height=59.4, units = "cm")
ggsave(path=(paste0(filepath,"figures")), filename = "Fig3.png", width=42, height=59.4, units = "cm", dpi=600) #remove later
```

# Drug-drug correlations
## Drug-drug correlation heatmap
```{r}

#output not as desired in R Markdown !!! -> check with Junyan (output fine in R script)

#prepare data for CLL
cll_df <- screen_long |> 
  filter(diagnosis == "CLL") |> 
  group_by(patientID, drug) |> 
  summarize(mean_viab = mean(viability)) |> 
  pivot_wider(names_from = drug, values_from = mean_viab)
  
cll_corr_df <- scale(cll_df[3:length(cll_df)])

#calculate correlations
cll_corr <- round(cor(cll_corr_df), 2)

#set colors, make correlations between -0.2 and 0.2 white
colors_cor_heatmap <- colorRampPalette(c(
  colorRampPalette(brewer.pal(9, "Spectral")[1:4])(7),  #for correlations below -0.2
  rep("white", 6),  # For correlations between -0.2 and 0.2
  colorRampPalette(brewer.pal(9, "Spectral")[6:9])(7)   #for correlations above 0.2
))

#add colors for pathways (names not correct!)
all_labels <- colnames(cll_corr)
highlight_hsp <- c("AMI.1", "AZD1208")
highlight_bcr <- c("BMS.509744", "Ceritinib", "Cisplatin", "Cobimetinib")
highlight_pi3k <- c("BGJ398", "BML.277")
highlight_aktmtor <- c("A.1210477", "A.674563", "Fludarabine", "Ganetespib")
highlight_mapk <- c("Cytarabine", "D4476", "Dabrafenib")

label_colors <- setNames(rep("black", length(all_labels)), all_labels)
label_colors[highlight_hsp] <- "#BEBADA"
label_colors[highlight_bcr] <- "#80B1D3"
label_colors[highlight_pi3k] <- "#FB8072"
label_colors[highlight_aktmtor] <- "#BC80BD"
label_colors[highlight_mapk] <- "#FDB462"

#plot
cll_cor_heatmap <- corrplot(cll_corr, type="lower", order="hclust", method="color", 
         col = colors_cor_heatmap(20), sig.level = 0.05, addgrid.col = "grey",
         insig = "blank", tl.col = label_colors, tl.cex = 0.7, tl.pos = "ld",
         tl.offset = c(0.3), cl.ratio = 0.1, title="CLL", cex.main = 0.75, col.main = "black", mar = c(0,0,0.5,0), 
         cl.cex = 0.7, cl.col = "grey", asp=1)
  
#prepare for multipanel figure
grid.echo()
P1 <- grid.grab()

#Plot1
matrix.colors1 <- getGrob(P1, gPath("square"), grep = TRUE)[["gp"]][["fill"]]
P1 <- editGrob(P1,
               gPath("square"), grep = TRUE,
               gp = gpar(col = NA,
                         fill = NA))

# apply the saved colours to the underlying matrix grob
P1 <- editGrob(P1,
               gPath("symbols-rect-1"), grep = TRUE,
               gp = gpar(fill = matrix.colors1))

# convert the background fill from white to transparent, while we are at it
P1 <- editGrob(P1,
               gPath("background"), grep = TRUE,
               gp = gpar(fill = NA))               
               

#prepare data for AML
aml_df <- screen_long |> 
  filter(diagnosis == "AML") |> 
  group_by(patientID, drug) |> 
  summarize(mean_viab = mean(viability)) |> 
  pivot_wider(names_from = drug, values_from = mean_viab)
  
aml_corr_df <- scale(aml_df[3:length(aml_df)])

#calculate correlations
aml_corr <- round(cor(aml_corr_df), 2)

#add colors for pathways (names not correct!)
all_labels <- colnames(aml_corr)
highlight_hsp <- c("Onalespib", "PF.3758309") # verify order
highlight_bcr <- c("KU.55933", "MI.503", "MK.2206", "Menin.MLL.Inhibitor")
highlight_pi3k <- c("SCH772984", "SGC.0946")
highlight_aktmtor <- c("Selisistat", "Sunitinib","Afatinib", "X10058.F4")
highlight_mapk <- c("PRT062607", "Palbociclib", "Panobinostat")

label_colors <- setNames(rep("black", length(all_labels)), all_labels)
label_colors[highlight_hsp] <- "#BEBADA"
label_colors[highlight_bcr] <- "#80B1D3"
label_colors[highlight_pi3k] <- "#FB8072"
label_colors[highlight_aktmtor] <- "#BC80BD"
label_colors[highlight_mapk] <- "#FDB462"

aml_cor_heatmap <- corrplot(aml_corr, type="lower", order="hclust", method="color", 
         col = colors_cor_heatmap(20), sig.level = 0.05, addgrid.col = "grey", 
         insig = "blank", tl.col = label_colors, tl.cex = 0.7, tl.pos = NULL,
         tl.offset = c(0.3), title="AML", cex.main = 0.75, col.main = "black", mar = c(0,0,0.5,0), 
         cl.ratio = 0.1, , cl.cex = 0.7, cl.col = "grey", asp=1)

#prepare for multipanel figure
grid.echo()
P2 <- grid.grab()

#Plot2
# save correlation matrix colors to a vector, then make coloured matrix grob transparent
matrix.colors2 <- getGrob(P2, gPath("square"), grep = TRUE)[["gp"]][["fill"]]
P2 <- editGrob(P2,
               gPath("square"), grep = TRUE,
               gp = gpar(col = NA,
                         fill = NA))

# apply the saved colours to the underlying matrix grob
P2 <- editGrob(P2,
               gPath("symbols-rect-1"), grep = TRUE,
               gp = gpar(fill = matrix.colors2))

# convert the background fill from white to transparent, while we are at it
P2 <- editGrob(P2,
               gPath("background"), grep = TRUE,
               gp = gpar(fill = NA))


#create a separate legend for pathways
colors_cor_heatmap <- df_pathway |> 
  filter(Pathway_mod %in% c("AKT/mTOR", "MAPK" , "B-cell receptor", "PI3K", "Stress pathway"))

colors_cor_heatmap$Pathway_mod[colors_cor_heatmap$Pathway_mod=="MAPK"] <- "MEK/ERK"
colors_cor_heatmap$random_value <- seq(1,nrow(colors_cor_heatmap),1)

legend_plot <- ggplot(colors_cor_heatmap, aes(x = Drug, y = random_value, fill = Pathway_mod)) +
  geom_tile() +
  scale_fill_manual(values = colors_cor_heatmap) +
  labs(fill = "Pathway") +
  theme_minimal() +
  theme(
  legend.title = element_text(size=8, face="bold"), 
  legend.text=element_text(size=8), legend.key.size = unit(0.3, 'cm'),
  legend.direction="vertical")+
  guides(fill = guide_legend(
    title.position="top", 
    title.hjust = 0,
    override.aes = list(shape = 1)))
  
legend <- get_legend(legend_plot)  
  
#put plots together
Fig4a <- plot_grid(P1, P2, ncol = 2, nrow = 1, align = "h", scale=1, rel_heights = c(1, 1), rel_widths = c(1, 1), greedy = TRUE)+
          draw_plot(legend, x=0.82, y=0.4, height=0.25, width=0.25)+
          theme(plot.margin = margin(t=1, unit = "cm"))
Fig4a

ggsave(path=(paste0(filepath,"figures")), filename = "Fig4a_correlation_heatmap.svg", width=42, height=30, units = "cm")

```

## Scatter plots for drug-drug correlations (Fig. 4B)
```{r, fig.width=12, fig.height=6}

#select drugs
drugs_show_corr <- c("SCH772984", "Cobimetinib", "Ibrutinib", "Idelalisib", "Cobimetinib", "Rapamycin")

#helper function for CLL
corr_drugs_cll <- screen_means |> 
  filter(diagnosis == c("CLL"), drug %in% drugs_show_corr) |> 
  pivot_wider(names_from = drug, values_from = mean_viab)

corr_drugs_aml <- screen_means |> 
  filter(diagnosis == c("AML"), drug %in% drugs_show_corr) |> 
  pivot_wider(names_from = drug, values_from = mean_viab)

corr_drugs <- rbind(corr_drugs_cll, corr_drugs_aml)

plot_corr <- function(data, x, y, xlab, ylab) {
  
  corr_labels <- data |> 
    group_by(diagnosis) |> 
    summarize(corr=cor(.data[[x]], .data[[y]], use = "complete.obs"),
              .groups = "drop")
  
  ggplot(data, aes_string(x = x, y = y)) +
    geom_point(aes(color = diagnosis), alpha = 0.5) +
    geom_smooth(method = "lm", se = TRUE,
                color = "black", fill = "lightgrey") + 
    labs(x = xlab, y = ylab, color="Diagnosis")+
    facet_wrap(~diagnosis)+
    theme_figures + 
    theme(strip.background = element_blank(),
          strip.text.x = element_blank())+
    geom_text(data = corr_labels, aes(x = 0.8, y = 1.1, 
                                      label = paste0("italic(R) == ", round(corr, 2))), size=3, parse = TRUE) +
    xlim(0.4,1.2) + ylim(0.4,1.2) +
    scale_color_manual(values = colors_diag)
}

#plot
p1 <- plot_corr(corr_drugs, "SCH772984", "Cobimetinib", xlab="SCH772984 (ERK)", ylab= "Cobimetinib (MEK)")
p2 <- plot_corr(corr_drugs, "Ibrutinib", "Idelalisib", xlab="Ibrutinib (BTK)", ylab= "Idelalisib (PI3K)")
p3 <- plot_corr(corr_drugs, "Ibrutinib", "Cobimetinib", xlab="Ibrutinib (BTK)", ylab= "Cobimetinib (MEK)")
p4 <- plot_corr(corr_drugs, "Rapamycin", "Idelalisib", xlab="Rapamycin (mTOR)", ylab= "Idelalisib (PI3K)")

Fig4b <- p1+plot_spacer()+p2+plot_spacer()+p3+plot_spacer()+p4+plot_layout(ncol=7, widths = c(4, 0.5, 4, 0.5, 4, 0.5, 4), guides = 'collect') & theme(legend.position='right', plot.margin = margin(t=1, l=0.5, unit = "cm"))

```

```{r, echo = FALSE}
#multipanel
Fig4 <- plot_grid(Fig4a,Fig4b, NULL, nrow = 3, rel_heights = c(0.45, 0.12,0.43), 
                  labels=c("A", "B", ""), label_size=14) & theme(plot.margin=margin(r=1,l=1,t=1,b=1,unit="cm"))

ggsave(path=(paste0(filepath,"figures")), filename = "Fig4.svg", width=42, height=59.4, units = "cm")
ggsave(path=(paste0(filepath,"figures")), filename = "Fig4.pdf", width=42, height=59.4, units = "cm")
ggsave(path=(paste0(filepath,"figures")), filename = "Fig4.png", width=42, height=59.4, units = "cm", dpi=600) #remove later

```

# Genotype-phenotype associations in CLL
## P value scatterplot
```{r}

#prepare data
#pTab <- ... (adapt from Junyans script)

pTab <- read.csv(paste0(filepath,"data/pTable_drug_VS_gene.csv"), sep=";") |> 
  mutate(across(c(p, p.adj, diff), ~as.numeric(gsub(",", ".", as.character(.)))))

#number of significant associations per gene (10% FDR)
plotTab1 <- filter(pTab, p.adj <= 0.1) %>% group_by(gene) %>%
  summarise(number = length(drug)) %>%
  arrange(desc(number)) %>% mutate(gene = factor(gene, levels = gene))

#p value scatter plot (IGHV not shown)
top_genes <- as.vector(unlist(plotTab1[2:8,"gene"]))
showGenes <- sort(top_genes)
append(showGenes, "other")

#define colors for each gene
library(RColorBrewer)
colorList <- c(colorRampPalette(brewer.pal(7, 'Set2'))(7), 
               "grey80", "grey90")
  
names(colorList) = c(showGenes, "other", "Below 10% FDR")

colorList["del17p"] <- "#FC8D62" 
colorList["del11q"] <- "#E5C494"
colorList["trisomy12"] <- "#FFD92F"
colorList["DDX3X"] <- "#E78AC3"
colorList["TP53"] <- "#66C2A5"

#define shapes for viability
shapeList <- ifelse(names(colorList) %in% showGenes, 19,1)
names(shapeList) <- names(colorList)
shapeList["other"] <- 19

shapeList <- c(1, 25, 24)
names(shapeList) = c("Below 10% FDR", "Down", "Up")

#order drugs by their association similarity
pMat <- dplyr::select(pTab, drug, gene, p) %>% 
  filter(gene!="IGHV.status") %>%
  spread(key = gene, value = p) %>%
  column_to_rownames("drug") %>% 
  as.matrix()

hc <- hclust(dist(pMat), method = "ward.D2")
drugOrder <- rownames(pMat)[hc$order]

#order drugs by p value
pval_list <- pTab |> 
  filter(gene != "IGHV.status") |> 
  group_by(drug) |> 
  summarize(pval = min(p)) |> 
  arrange(pval, descending = FALSE)

drugOrder2 <- pval_list$drug

#prepare plot table
plotTab <- pTab %>% filter(gene != "IGHV.status") %>%
  mutate(gene = ifelse(gene %in% showGenes, gene, "other")) %>%
  mutate(gene_FDR = ifelse(p.adj < 0.1, gene, "Below 10% FDR"),
         drug = factor(drug, levels = drugOrder2),
         gene = factor(gene, levels = names(colorList)),
         direction = ifelse(gene_FDR == "Below 10% FDR", "Below 10% FDR", ifelse(sign(diff)>0, "Down", "Up"))) 

#plot
Fig5a <- ggplot(plotTab, aes(x=drug, y = -log10(p), color = gene, shape = direction)) +
  geom_point(size =2, alpha = 0.8, stroke = 1) + scale_color_manual(values = colorList) +
  scale_shape_manual(values = shapeList, name = "Viability effect") + theme_bw() +
  labs(x="", y = expression(-log[10]*italic(P)), color = "Genetic aberration")+
  theme(axis.text.x = element_text(angle = 45, hjust =1, vjust=1, 
                                   color="black", size=8), 
        axis.text.y = element_text(color="black", size=8), 
        axis.title.y = element_text(size=8), 
        legend.title = element_text(size=8, face="bold"), 
        legend.text=element_text(size=8),
        panel.grid.major.x = element_line(size=.1, color="grey50", 
                                          linetype = "dotted"),
        panel.grid.major.y = element_blank(),
        panel.grid.minor.y = element_blank(),
        panel.border = element_rect(color = 'black', 
                                    fill = NA, 
                                    size = 1),
        plot.margin = margin(t=1, l=1, unit = "cm"),
        legend.position = c(.95, .95),
        legend.justification = c("right", "top"),
        legend.box.just = "right",
        legend.box = "horizontal", 
        legend.margin = margin(6, 6, 6, 6))+
  geom_hline(yintercept=-log10(0.004706046), linetype = "dashed", linewidth=0.3, color="black")+
  annotate("text", label="FDR 10%", size=3, x=60, y=3)+
  guides(shape = guide_legend(order = 1),
         color = guide_legend(order = 2))
Fig5a

ggsave(path=(paste0(filepath,"figures")), filename = "Fig5a_pvalue_scatterplot.svg")

```

## Volcano plots (for IGHV, trisomy12 and TP53)
```{r}


```


## Dose-response relationship in DDX3X-mutated U-CLL

### BTK vs PI3K

## Lasso model

## del11q

## trisomy(12) in M-CLL and U-CLL

## Fludarabine and nutlin-3a in TP53-mutated CLL


# DDX3X
## Overview of mutations
### Table

### Lollipop plot

### RNA-seq gene set signature

### RNA seq volcano plot for single genes

### Heatmaps

### Genes of interest: NLRP3 and DDX3Y


# Longitudinal analysis
## Distribution

## Treatment contexts

## Longitudinal changes in drug response

### P value heatmap

### Drugs of interest

### Absolute differences in mean viability


# Clinical correlation...

```
